units/jasper_deploy.go

```diff
@@ -0,0 +1,405 @@
+	j := makeJasperDeployJob()
+	j.env = env
+	j.HostID = h.Id
+	j.host = &h
+	j.CredentialsExpiration = expiration
+	j.DeployThroughJasper = deployThroughJasper
+	jobID := fmt.Sprintf("%s.%s.%s", jasperDeployJobName, j.HostID, id)
+	if deployThroughJasper {
+		jobID += ".deploy-through-jasper"
+	}
+	j.SetID(jobID)
+	return j
+}
+
+// Run deploys new Jasper credentials to a host.
+func (j *jasperDeployJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if err := j.populateIfUnset(ctx); err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message": "could not populate required fields",
+			"host":    j.HostID,
+			"job":     j.ID(),
+		}))
+		j.AddError(err)
+		return
+	}
+
+	defer func() {
+		if j.HasErrors() {
+			if err := j.tryRequeueDeploy(ctx); err != nil {
+				grip.Error(message.WrapError(err, message.Fields{
+					"message":  "could not requeue Jasper deploy job",
+					"host":     j.host.Distro.Id,
+					"distro:":  j.host.Distro.Id,
+					"expires":  j.CredentialsExpiration,
+					"attempts": j.host.JasperDeployAttempts,
+					"job":      j.ID(),
+				}))
+				return
+			}
+		}
+	}()
+
+	settings := j.env.Settings()
+
+	// Update LCT to prevent other jobs from trying to terminate this host,
+	// which is about to kill the agent.
+	if err := j.host.UpdateLastCommunicated(); err != nil {
+		j.AddError(errors.Wrapf(err, "error setting LCT on host %s", j.host.Id))
+	}
+
+	grip.Info(message.Fields{
+		"message": "deploying Jasper service to host",
+		"host":    j.host.Id,
+		"distro":  j.host.Distro.Id,
+		"job":     j.ID(),
+	})
+
+	creds, err := j.host.GenerateJasperCredentials(ctx, j.env)
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message": "problem generating new Jasper credentials",
+			"host":    j.host.Id,
+			"distro":  j.host.Distro.Id,
+			"job":     j.ID(),
+		}))
+		j.AddError(err)
+		return
+	}
+
+	writeCredentialsCmd, err := j.host.WriteJasperCredentialsFileCommand(creds)
+	if err != nil {
+		grip.Error(message.Fields{
+			"message": "could not build command to write Jasper credentials file",
+			"host":    j.host.Id,
+			"distro":  j.host.Distro.Id,
+			"job":     j.ID(),
+		})
+		grip.Error(err)
+		j.AddError(err)
+		return
+	}
+
+	if j.DeployThroughJasper {
+		client, err := j.host.JasperClient(ctx, j.env)
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": "could not get Jasper client",
+				"host":    j.host.Id,
+				"distro":  j.host.Distro.Id,
+				"job":     j.ID(),
+			}))
+			j.AddError(err)
+			return
+		}
+		// We use this ID to later verify the current running Jasper service.
+		// When Jasper is redeployed, its ID should be different to indicate it
+		// is a new Jasper service.
+		serviceID := client.ID()
+
+		writeCredentialsOpts := &jasper.CreateOptions{
+			Args: []string{
+				"bash", "-c", writeCredentialsCmd,
```

Comment:
why do we need to run this in batch? 

how do we ensure that the string is properly quoted?

Reply:
I think you can specify an io.Rearder to the stdin of a processs for execing, which should work?

Reply:
redirect might be harder to fake out.

Reply:
The bytes of the io.Reader won't be serialized when sending the CreateOptions over RPC/SSH.

Reply:
I made [a ticket](https://jira.mongodb.org/browse/MAKE-865) to handle standard input for remote processes.